### ISSUE TITLE
@swc/core threw an error when attempting to validate swc compiler options.

### ISSUE THREAD

**Issue by seb-mtl** [swc-project/swc#5616]
### Describe the bug

Hi there,

I am using `@swc/core` to speedup the execution of my test suite with [avajs/ava](https://github.com/avajs/ava).

**System info:**
    Ubuntu 20.04.1 (VM via parallels 18 on a Mac Mini Intel)
    node 16.15.0

During `npm run test` I receive the following error, and I am not sure how to resolve the issue.

```typescript
/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/transpilers/swc.ts:262
      throw new Error(
            ^
Error: @swc/core threw an error when attempting to validate swc compiler options.
You may be using an old version of swc which does not support the options used by ts-node.
Try upgrading to the latest version of swc.
Error message from swc:
Bindings not found
    at createVariant (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/transpilers/swc.ts:262:13)
    at createSwcOptions (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/transpilers/swc.ts:211:25)
    at create (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/transpilers/swc.ts:56:41)
    at createTranspiler (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/index.ts:784:16)
    at createTranspileOnlyGetOutputFunction (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/index.ts:1341:28)
    at createFromPreloadedConfig (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/index.ts:1404:34)
    at create (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/index.ts:624:10)
    at register (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/index.ts:591:15)
    at Object.registerAndCreateEsmHooks (/home/st/actions-runner/_work/snow/snow/node_modules/ts-node/src/esm.ts:116:34)
    at file:///home/st/actions-runner/_work/snow/snow/node_modules/ts-node/esm.mjs:8:7
  ✖ test/1.sys.test.ts exited with a non-zero exit code: 1
```

### Config

Snippet of `tsconfig.json` config:

```json
  "ts-node": {
    "swc": true
  },
```

### Expected behavior

Successful execution of `npm run test`.

### Version

v1.2.241

**Comment by kdy1** [comment]
Which os/cpu are you using?

**Comment by kdy1** [comment]
Also, did you use something like `--skip-optional` or `--ignore-scripts`?

**Comment by seb-mtl** [comment]
I added more information under **System Info**.  No `--skip-optional` nor `--ignore-scripts` is used.

Also for completeness, here is the definition of `npm run test" in `package.json`:
```
"test": "nyc --reporter=html --reporter=lcov --reporter=text ava"
```

**Comment by seb-mtl** [comment]
Thanks for the quick reply! I added more information under **System Info**.  No `--skip-optional` nor `--ignore-scripts` is used.

Also for completeness, here is the definition of `npm run test` in `package.json`:
```
"test": "nyc --reporter=html --reporter=lcov --reporter=text ava"
```

**Comment by seb-mtl** [comment]
Quick update, the same project successfully built on the same machine (also Parallels) under Windows. Updating to `swc/core@1.42.0`works now also under Linux. Could the error come back or did the latest version fix something that could explain the issue on Linux?

**Comment by kdy1** [comment]
Did you share the project directory? You should install on target platform unless you are overriding npm config

**Comment by ejkg** [comment]
I was able to reproduce this issue with my Apollo Federation Gateway running in a Node docker container. I can provide the Dockerfile

```Dockerfile
FROM node:17.3
WORKDIR /app
ENV NODE_ENV=dev
ENV ZIPKIN=http://localhost:9411
EXPOSE 4000
COPY . .
CMD ["yarn", "run", "dev"]
```

encountered this error on `docker run my-image:1.0` 

```
$ docker run graphql-federation-poc:0.2
yarn run v1.22.17
$ ts-node src/index.ts
/app/node_modules/ts-node/src/transpilers/swc.ts:262
      throw new Error(
            ^
Error: @swc/core threw an error when attempting to validate swc compiler options.
You may be using an old version of swc which does not support the options used by ts-node.
Try upgrading to the latest version of swc.
Error message from swc:
Bindings not found
    at createVariant (/app/node_modules/ts-node/src/transpilers/swc.ts:262:13)
    at createSwcOptions (/app/node_modules/ts-node/src/transpilers/swc.ts:211:25)
    at create (/app/node_modules/ts-node/src/transpilers/swc.ts:56:41)
    at createTranspiler (/app/node_modules/ts-node/src/index.ts:784:16)
    at createTranspileOnlyGetOutputFunction (/app/node_modules/ts-node/src/index.ts:1341:28)
    at createFromPreloadedConfig (/app/node_modules/ts-node/src/index.ts:1404:34)
    at phase4 (/app/node_modules/ts-node/src/bin.ts:543:44)
    at bootstrap (/app/node_modules/ts-node/src/bin.ts:95:10)
    at main (/app/node_modules/ts-node/src/bin.ts:55:10)
    at Object.<anonymous> (/app/node_modules/ts-node/src/bin.ts:800:3)
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 1.
```

**Comment by Bluzzi** [comment]
I have the same problem, I can't start my project at all with Docker/TS-Node/SWC

**Comment by RiESAEX** [comment]
Can you try to reinstall swc in your docker?

**Comment by Bluzzi** [comment]
That's what I am doing

Dockerfile:
```docker
FROM node:16.16-alpine

WORKDIR /app

COPY package*.json ./
RUN npm install
COPY . .

CMD npm run start
```

.dockerignore:
```docker
node_modules
```

**Comment by Bluzzi** [comment]
This issue is still valid

**Comment by btecu** [comment]
I'm also running into this issue after updating from `node 14.17.1` to `node 16.16.0` on a MBP 16 M1 Pro:
```
/Users/.../node_modules/@swc/core/index.js:250
        throw new Error("Bindings not found");
        ^

Error: Bindings not found
    at Compiler.transformSync (/Users/.../node_modules/@swc/core/index.js:250:15)
    at transformSync (/Users/.../node_modules/@swc/core/index.js:348:21)
    at transformSync (/Users/.../node_modules/@swc-node/core/index.ts:69:26)
    at compile (/Users/.../node_modules/@swc-node/register/register.ts:44:40)
    at exts (/Users/.../node_modules/@swc-node/register/register.ts:55:38)
    at Module._compile (/Users/.../node_modules/pirates/lib/index.js:130:29)
    at Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Object.newLoader [as .ts] (/Users/.../node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
```

**Comment by btecu** [comment]
The issue is still valid.

**Comment by kwonoj** [comment]
Not sure why bot aggressively marks this one 🤔

**Comment by btecu** [comment]
Thanks @jgujgu. Is there anything for `npm`?

**Comment by gspetrou** [comment]
Getting this error on an M1 Mac using node 19, but not on 18.

**Comment by jcampbell05** [comment]
I'm getting this issue with the Node 16 docker image.

**Comment by wodCZ** [comment]
I've ran into similar issue with a bit uncommon setup, but maybe this will help shed some light.

I have a working setup on local node 18, using yarn. I wanted to create a sample Dockerfile using npm, so I ran `npm i` locally to generate `package-lock.json`. Then docker build ran up to the build point where it failed on `Fallback bindings does not support filesystem access.`.
When installing `@swc/core` and `@swc-cli` in the build step in the container again, it worked well.

Then I removed `node_modules` and `package-lock.json` on the host and ran `npm i` again to get a new `package-lock.json`. And with this new lock the whole build ran successfully.

[old-package-lock.json.txt](https://github.com/swc-project/swc/files/9982277/old-package-lock.json.txt)
[package-lock.json.txt](https://github.com/swc-project/swc/files/9982276/package-lock.json.txt)

When compared, there are few `@swc` packages that didn't get "locked" with existing node_modules.

So, maybe the "work-around" is to `rm -rf node_modules && rm package-lock.json && npm i` on the host?

**Comment by jcampbell05** [comment]
My issue was because I was mounting a volume in such a way the node_modules would be mounted to the host which could cause issues.

By adding node_modules in it's own volume in the docker-compose spec I can tell docker to preserve the node_modules folder from the underlying image and now it works.

**Comment by ZuBB** [comment]
Hi there!

I also had this issue and was able to resolve it. below are my "story".

Most beneficial (give me some food to think) to me were messages from @jgujgu and @wodCZ.

I was migrating from `yarn` to `npm`, and I wanted to have a dependency tree as similar as possible. So I did not delete `node_modules` folder and just run `npm i`. This caused `package-lock.json` file to be not fully valid, where swc package for my arch was missing (I figured out that later). So I did a couple of test setups, and they were OK. At some point, I decided to check "moving parts" of the swc software in the file system and found that arm64 file/package was missing.

----

**Summary**: if you got a subject error message -- make sure you have both arch package and arch binary in your FS

**Advice to developers**: maybe you should add/adjust a check if arch binary exists and make sure the error message for that is crystal clear

**Comment by Trojan0523** [comment]
using node v16.18.1 also met this problem, on MBP 13 M1, problem performed like @btecu 
> Error: Bindings not found
    at Compiler.transformSync (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/@swc+core@1.3.20/node_modules/@swc/core/index.js:250:15)
    at Object.transformSync (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/@swc+core@1.3.20/node_modules/@swc/core/index.js:348:21)
    at compile (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/@swc+register@0.1.10_@swc+core@1.3.20/node_modules/@swc/register/lib/node.js:76:22)
    at compileHook (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/@swc+register@0.1.10_@swc+core@1.3.20/node_modules/@swc/register/lib/node.js:91:16)
    at Module._compile (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/pirates@4.0.5/node_modules/pirates/lib/index.js:130:29)
    at Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Object.newLoader [as .js] (/Users/buxiongyu/Desktop/zeus/node_modules/.pnpm/pirates@4.0.5/node_modules/pirates/lib/index.js:141:7)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Module.require (node:internal/modules/cjs/loader:1057:19)


i'm just wondering if it is cross platform problem of npm/pnpm ?

**Comment by DzmVasileusky** [comment]
Same issue on 
```
Node : 16.15.0
   OS   : darwin x64
   npm  : 8.10.0
```

Experience this while using `@nrwl/nx` npm package.

**Comment by DzmVasileusky** [comment]
What have helped me
1. Create new clean project
2. manually install `@swc/core` of needed version there
3. manually copypaste `core-darwin-x64` to `node_modules/@swc`

**Comment by nephix** [comment]
What helped me in my npm monorepo:

1. `npm uninstall` all `swc` related packages
2. remove `package-lock.json`
3. remove all `node_modules` directories
4. `npm install`
5. `npm install` the `swc` related packages again

**Comment by roftLin** [comment]
Hey buddy, I have encountered the same issue as you and have solved it, I would like to share my experience with you to see if it can help.
Just a few days ago I was able to install Taro-cli without any issues, then after reinstalling the system, my project encountered the same error as yours. I searched for solutions in the Taro repository and the SWC repository, but there was no ready answer. However, they gave some directions to troubleshoot.So I tried switching the node version, package manager version and source, and Taro/cli version, but all failed. Finally, I debugged the source code with breakpoints, combined with my own experience, and found that the problem was related to the rust runtime environment, which needed to be installed. The specific installation method can be referred to the link below.
https://blog.csdn.net/u011149152/article/details/123518465

**Comment by littleski** [comment]
I think there is a bug in the postinstall.sh script in the swc module. Somewhere, it "trash" the @swc/wasm module instead of adding it.
the script however is discarded if @swc/wasm is added in the dependencies. What fixed it for me (in my dockers in CI), is to add the package in the dependencies.

**Comment by arxpoetica** [comment]
@littleski I ran into the same problem and adding `@swc/wasm` solved it for me too...so yes, a bug in the post install maybe?

**Comment by pbredenberg** [comment]
I ran into a similar problem when running a build with SWC in GitHub actions (worked fine on my M1 mac). 

The build would fail with `Fallback bindings does not support filesystem access.`

After upgrading @swc/cli to 0.1.62 and @swc/core to 1.3.39 the build executed again in my workflow.

**Comment by Squashhh** [comment]
Does anyone have a solution?

**Comment by TIMMLOPK** [comment]
I have the same issue when I use Github action on @swc/core 1.3.44. I try to downgrade to @swc/core 1.3.39.The action work again.

**Comment by aleczratiu** [comment]
> What have helped me
> 
> 1. Create new clean project
> 2. manually install `@swc/core` of needed version there
> 3. manually copypaste `core-darwin-x64` to `node_modules/@swc`

Indeed this solution is working well, btw installing `@swc/core` is enough and solves the problem, but I'm Interested if there are other dependencies that require it.

**Comment by 2b1q** [comment]
M1, node v18.12.1, pnpm (solved)
before I added  `@swc/core-linux-arm64-musl` as dev deps, had same err 
```
  ● Test suite failed to run
    Bindings not found
      at Compiler.transformSync (node_modules/.pnpm/@swc+core@1.3.40/node_modules/@swc/core/index.js:250:15)
```

**Comment by mmatila** [comment]
Having similar issues here.

M1 Pro
`node v19.7.0` (tried various previous versions as well)
`@swc/jest ^0.2.26`

```
Error: Jest: Got error running globalSetup - /app/tests/globalSetup.ts, reason: Bindings not found

at Compiler.transformSync (/node_modules/@swc/core/index.js:250:15)
```

Tried going through the whole install/uninstall process with `swc` packages and `node_modules` that seemed to work for some people in the thread – not for me.

**Comment by DzmVasileusky** [comment]
this issue is annoying I have to install swc manually each time after `npm i`, does anybody found a way to fix it once for all?

**Comment by kwonoj** [comment]
Primarily this is due to you have installed one binaries for the specific platform and tries to run on other platform which doesn't have corresponding binary. Upstream issues like https://github.com/npm/cli/issues/4828 make things more challenging. 

We are open to accept fixes, but I don't believe there's no quick win for now.

**Comment by DzmVasileusky** [comment]
@kwonoj well, we are sharing project among people with MacOS, Windows, Ubuntu but this is a normal situation when repo has collaborators

**Comment by kwonoj** [comment]
I'm not saying it's uncommon to share project across different platform users, saying there are some constraint of actions we can take. As said, we are open to accept fix / suggestions but afaik there is no clear win to any paths we can take.

**Comment by DaSchTour** [comment]
I think that the approach how bindings are handled is not compatible to `package-lock.json` and `npm ci`.
When installing (generating package-lock) on one OS and running the CI task on another OS by using `npm ci` this will always fail because the wrong bindings are persisted in the package-lock.

Or maybe it's because `swc` doesn't use the expected postinstall scripts or so that would work with the recommended approach when using private repositories on a CI like described in this documentation: https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#use-private-packages

In short when using `npm ci --ignore-scripts` for security reasons `swc` doesn't setup properly when running `npm rebuild && npm run prepare --if-present` afterwards.

**Comment by DzmVasileusky** [comment]
We switched to `pnpm`, it is not only faster it also doesn't have this issue as well as issues with legacy peer dependencies in lock file

**Comment by moltar** [comment]
Having the same issue. The only solution is to completely nuke the nvm (fnm) Node environment, and the local node_modules and do a full re-install. Seems like something gets written into the global scope, that has an effect on this?

**Comment by moltar** [comment]
Never mind, just did it again without node (nvm) nuking, just `node_modules` and it worked.

For us, the cause of this misalignment, I think, is Renovate Bot. Every time it updates a dep, and I do a pull, then the breakage will occur.

**Comment by SkyaTura** [comment]
Same problem. Runs fine on MacOS in local env and throws while building on Docker

**Comment by SkyaTura** [comment]
`pnpm` solved the build on CI, but my coworker is unable to run the project on Ubuntu (Node v18.16.0)

**Comment by jaybe78** [comment]
Getting same issue. pnpm does not work for me

**Comment by jaybe78** [comment]
<img width="1032" alt="Screenshot 2023-07-25 at 09 34 33" src="https://github.com/swc-project/swc/assets/3605098/4b195462-96ca-4ce1-bb69-c0886d851625">

What the hell ?

**Comment by kdy1** [comment]
@jaybe78 https://kdy1.dev/posts/2023/1/plus-1-is-spam

**Comment by kdy1** [comment]
@jaybe78 This is not a bug of swc

**Comment by littleski** [comment]
> I think there is a bug in the postinstall.sh script in the swc module. Somewhere, it "trash" the @swc/wasm module instead of adding it. the script however is discarded if @swc/wasm is added in the dependencies. What fixed it for me (in my dockers in CI), is to add the package in the dependencies. But looking at the comments, i think there is multiple issues listed here, so it may not fix for everyone.

I put it again here, in case it helps someone.

**Comment by pcdevil** [comment]
adding the core modules as optional dependencies for the project fixed the error for me:

```bash
npm install --save-optional \
    "@swc/core-darwin-arm64" \
    "@swc/core-darwin-x64" \
    "@swc/core-linux-arm-gnueabihf" \
    "@swc/core-linux-arm64-gnu" \
    "@swc/core-linux-arm64-musl" \
    "@swc/core-linux-x64-gnu" \
    "@swc/core-linux-x64-musl" \
    "@swc/core-win32-arm64-msvc" \
    "@swc/core-win32-ia32-msvc" \
    "@swc/core-win32-x64-msvc"
```

you can also filter out the list of packages in case you are certain what is the OS / architecture of your development computer and the CI.

hope this helps for others until the issue is resolved!

**Comment by kanocarra** [comment]
I was having the same issue `Fallback bindings does not support filesystem access.` when running in GitHub Actions. 

@pcdevil 's response solved the issue - thanks!!

**Comment by bodinsamuel** [comment]
thanks @pcdevil that was the solution for me (when compiling react-email in docker)

**Comment by svargas-dev** [comment]
I'm having this issue re-appear with `@swc/core@1.3.83` on Linux (MX Linux 23 on x64 -- Debian 12 based) with a mono repo using npm. If I explicitly install `@swc/core@1.3.83` in my workspace it doesn't work but explicitly installing the previous version `npm -w=backend i -D @swc/core@1.3.82` is a good workaround for now.

**Comment by corazon-roto** [comment]
I can work around this by simply `yarn add -D @swc/core@latest`.

**Comment by MohitOru** [comment]
just clear `node_modules` files and do `npm install`

**Comment by jwalton** [comment]
I worked around this with:

```
npm install --save-dev @swc/core@1.3.2 @swc/cli@0.1.57
npm install @swc/helpers@0.4.11
```

These are pretty old versions - there's likely newer ones that work, I just grabbed these from a project where I didn't have this issue.

**Comment by throrin19** [comment]
I'm having same issue using `@swc/core:1.5.2`

**Comment by tfeltwell** [comment]
@throrin19 there is a deprecation warning against `1.5.2` now: `@swc/core@1.5.2: Mac OS installation is broken`. Upgrading to `1.5.3` fixes the issue.

**Comment by kdy1** [comment]
Can someone provide a minimal repro?

**Comment by bediu** [comment]
@kdy1 

The installation always fails on certain servers. For me specifically, in the cheap ones running on AMD vCPUs.

`uname -m` output is `aarch64`

`lscpu` output is: 
```
Architecture:           aarch64
  CPU op-mode(s):       32-bit, 64-bit
  Byte Order:           Little Endian
CPU(s):                 4
  On-line CPU(s) list:  0-3
Vendor ID:              ARM
  Model name:           Neoverse-N1
    Model:              1
    Thread(s) per core: 1
    Core(s) per socket: 4
    Socket(s):          1
    Stepping:           r3p1
    BogoMIPS:           50.00
    Flags:              fp asimd evtstrm aes pmull sha1 sha2 crc32 atomics fphp asimdhp cpuid asimdrdm lrcpc dcpop asimddp ssbs

```
Let me know if I can provide you with nay additional information.

(We have resorted to version locking swc manually in our monorepo for now, which isn't good. )

**Comment by nparoski** [comment]
For those still struggling heres what worked for me

> I've found a fix, apparently what happens (not sure why) is node_modules/@swc/core installs wrong optional packages within docker container.
> 
> Heres the fix:
> 
> Open docker image and find your WORKDIR
> Remove core-darwin-x64 or core-win-32-x64-msvc or similar from node_modules/@swc/core.
> There should be core-linux-x64-gnu and core-linux-x64-muscl (if your docker image is linux)
> Run npm install from within docker container (this time you should see correct packages in @swc/core under node_modules/)
> My understanding is that when you pull project and do npm install on host machine which is macOS or Windows, swc installs right packages for your host system.
> 
> I'm not sure why it skips installing correct ones on docker image but this is what causes issues.
> I tried setting this up on windows and macbook and both times had the same issues where this was the fix.

https://github.com/orgs/adonisjs/discussions/4600

**Comment by LeonardoRochaInacio** [comment]
If your problem is locally, I mean, you are using docker-compose pay attention to the shared volumes, problem it is sharing the wrong file between your windows system and your container running windows. 

You can solve it excluding your node_modules from the sharing volumes from you docker-compose.yml file like this:

```
volumes:
      - ./:/home/node/app # Here we are sharing all the content from local to your container
      - /home/node/app/node_modules/ # Here we are excluding the node_modules
```
